### PR TITLE
Update Trilinos to C++14

### LINKED ISSUE #6260: Trilinos: Transition Trilinos to require c++14 standard
Transition Trilinos to require c++14 standard rather than c++11. Many apps are already building with c++14 and Kokkos would benefit greatly from this. This ticket is to outline requirements and start discussions with stakeholders. 

Compiler support:
Aggregate table of compiler support: https://en.cppreference.com/w/cpp/compiler_support

Minimum compiler version that has full support for c++14:
- gcc: 5.0 (https://gcc.gnu.org/projects/cxx-status.html#cxx14)
- clang: 3.4 (https://clang.llvm.org/cxx_status.html)
- intel: 17.0 
   - supports everything except N3664: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3664.html
- ibm xl c/c++: 16.1 
- Visual Studio: VS 2017, 15.7 (https://docs.microsoft.com/en-us/cpp/overview/visual-cpp-language-conformance?view=vs-2019)

From a SEMS standpoint, it looks like we are in good shape.

To do:
* [ ] Discussion with SART teams. Any blockers?
* [ ] Check on SEMS module support for all compilers
* [ ] Set up test builds 
* [ ] Switch PR testing to c++14 
* [ ] ?

@jwillenbring @bartlettroscoe @jmgate @crtrott @mwglass @maherou @prwolfe @trilinos/framework

### PR DESCRIPTION
**Pull request by bartlettroscoe**
This updates Trilinos to require C++14 by default.   See updated documentation for details.

Closes #6260.

## How was this tested?

I tested this locally manually with several different use cases:

* Default configure (don't set `CMAKE_CXX_STANDARD`) => Produces `CMAKE_CXX_STANDARD:STRING=14` in the `CMakeCache.txt` file and produces the compiler flag `-std=c++14` with GCC 7.2.0.
* Set `-D CMAKE_CXX_STANDARD=11` => Errors out right away with the error message `CMAKE_CXX_STANDARD=11 is not in the allowed set (14|17|20)`
* Set `-D CMAKE_CXX_STANDARD=14` => Configures the same as default configure
* Set `-D CMAKE_CXX_STANDARD=17` => Results in compiler flag `-std=c++1z` with GCC 7.2.0 (which is the same thing as  `-std=c++17` according to GCC documentation).

## Checklist

* [X] Merge this branch to the 'atdm-nighlty-manual-updates' branch and push to get this tested with the ATDM Trilinos builds.
* [X] Revert the merge of the merge commit to the 'atdm-nighlty-manual-updates' branch and push.
* [ ] Reproduce and fix the build failures in the 'tlcc2' and 'ats2' builds shown [below](https://github.com/trilinos/Trilinos/pull/8317#issuecomment-723138268).
* [ ] Revert the revert commit e2ebb2f9addd969af3cfe2823776d66efa4da169 on the 'atdm-nightly-manual-updates' branch to get this tested again in the ATDM Trilinos builds.
* [ ] After getting back passing results from the ATDM Trilinos builds (except for perhaps a few new test failures), merge this branch to the 'develop' branch.
* [ ] Create new ATDM Trilinos GitHub issues for any new failures caused by this change.

### CODE REVIEW

**Comment by bartlettroscoe** [comment]
FYI: I merged this to the 'atdm-nightly-manual-updates' branch so that it will get tested on all of the ATDM Trilinos builds (except for the ATS-1 builds on 'mutrino' which are currently not running, see [ATDV-405](https://sems-atlassian-srn.sandia.gov/browse/ATDV-405).

**Review by rppawlo — DISMISSED** [review]
Thanks @bartlettroscoe !

**Comment by bartlettroscoe** [comment]
@kddevin, @etphipp, @rppawlo, @trilinos/framework, @trilinos/kokkos,

I merged this branch to the 'atdm-nightly-manual-updates' branch last night and today we are seeing some new build errors in the ATDM Trilinos builds as you can see at:

* https://testing.sandia.gov/cdash/index.php?project=Trilinos&date=2020-11-06&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=65&value1=Trilinos-atdm-&field2=builderrors&compare2=43&value2=0

which shows:

| Site | Build Name | Update | Update Time | Conf Err | Conf Warn | Conf Time | Build Err | Build Warn | Build Time | Test Not Run | Test Fail | Test Pass | Test Time | Test Proc Time | Start Test Time | Labels |
| -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- |
vortex | Trilinos-atdm-ats2-cuda-10.1.243-xl-2020.03.18_spmpi-rolling_static_dbg | 55906c | 2m 18s | 0 | 4 | 3m 52s | 20 | 0 | 5m 20s | 0 | 0 | 0 | 0s | 0s | 2 hours ago | (16 labels)
chama | Trilinos-atdm-tlcc2-intel-debug-openmp | 55906c | 2m 18s | 0 | 19 | 7m 9s | 50 | 0 | 56m 58s | 0 | 0 | 2133 | 44m 52s | 5h 46m 49s | 3 hours ago | (29 labels)
vortex | Trilinos-atdm-ats2-cuda-10.1.243-xl-2020.03.18_spmpi-rolling_static_opt | 55906c | 1m 12s | 0 | 4 | 2m 54s | 20 | 0 | 3m 5s | 0 | 0 | 0 | 0s | 0s | 3 hours ago | (16 labels)
skybridge | Trilinos-atdm-tlcc2-intel-opt-openmp | 55906c | 2m 18s | 0 | 19 | 6m 14s | 50 | 0 | 49m 20s | 0 | 0 | 2133 | 16m 41s | 2h 10m 36s | 4 hours ago | (29 labels)

and 3 of the 4 of these builds show Kokkos build errors as shown at:

* https://testing.sandia.gov/cdash/index.php?project=Trilinos&date=2020-11-06&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=65&value1=Trilinos-atdm-&field2=builderrors&compare2=43&value2=0&field3=subprojects&compare3=93&value3=Kokkos

showing:

| Site | Build Name | Update | Update Time | Conf Err | Conf Warn | Conf Time | Build Err | Build Warn | Build Time | Test Not Run | Test Fail | Test Pass | Test Time | Test Proc Time | Start Test Time | Labels |
| -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- |
vortex | Trilinos-atdm-ats2-cuda-10.1.243-xl-2020.03.18_spmpi-rolling_static_dbg | 55906c | 2m 18s | 0 | 4 | 3m 52s | 20 | 0 | 5m 20s | 0 | 0 | 0 | 0s | 0s | 2 hours ago | Kokkos
chama | Trilinos-atdm-tlcc2-intel-debug-openmp | 55906c | 2m 18s | 0 | 19 | 7m 9s | 6 | 0 | 56m 58s | 0 | 0 | 32 | 44m 52s | 4m 56s | 3 hours ago | Kokkos
vortex | Trilinos-atdm-ats2-cuda-10.1.243-xl-2020.03.18_spmpi-rolling_static_opt | 55906c | 1m 12s | 0 | 4 | 2m 54s | 20 | 0 | 3m 5s | 0 | 0 | 0 | 0s | 0s | 3 hours ago | Kokkos

I am going to go ahead and revert the merge of this branch from the 'atdm-nightly' branch then someone will need to reproduce these issues and fix them offline (so they don't damage the ATDM Trilinos builds).

Please do not merge this branch until someone can fix all of these build errors.

**Comment by bartlettroscoe** [comment]
FYI: I just pushed the revert commite2ebb2f9addd969af3cfe2823776d66efa4da169 to the 'atdm-nightly-manual-updates' branch.  Therefore, tomorrow, this change will be gone from the ATDM Trilinos nightly builds.

I will put in a checklist of things that must be done before merging this PR and turning on C++14 by default in all Trilinos builds.

**Comment by ndellingwood** [comment]
@bartlettroscoe 

> and 3 of the 4 of these builds show Kokkos build errors as shown at:

>    https://testing.sandia.gov/cdash/index.php?project=Trilinos&date=2020-11-06&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=65&value1=Trilinos-atdm-&field2=builderrors&compare2=43&value2=0&field3=subprojects&compare3=93&value3=Kokkos

I looked at some of these build errors, one problem is that the loaded gcc version (4.9.3) is too old and no longer supported with the switch to c++14, e.g.

```
nvcc warning : The -std=c++14 flag is not supported with the configured host compiler. Flag will be ignored.
In file included from /vscratch1/projects/usr/tce/packages/gcc/gcc-4.9.3/gnu/include4.9.3/cstdint:35:0,
```

Can the `environment.sh` for these systems be updated to swap out the gcc/4.9.3 module for a supported version 5.3 or newer?

**Comment by bartlettroscoe** [comment]
> I looked at some of these build errors, one problem is that the loaded gcc version (4.9.3 on vortex) is too old and no longer supported with the switch to c++14

@ndellingwood, you refer to the 'ats2-cuda-10.1.243-xl-2020.03.18_spmpi-rolling' builds on 'vortex'?

> Can the environment.sh for these systems be updated to swap out the gcc/4.9.3 module for a supported version 5.3 or newer?

The envs on the 'ats2' system 'vortex' is controlled by SPARC and I suspect that might be the officially supported GCC version for that XL compiler on ATS-2.  I will put you in contact with the SPARC developers offline to see if you guys can figure this out.  

Otherwise, we may need to hold off on upgrading Trilinos to require C++14 until after the completion of the L1 milestone.  (Then this we become an issue for the @trilinos/framework team and the Trilinos product leads to address.)

**Comment by ndellingwood** [comment]
@bartlettroscoe 

> @ndellingwood, you refer to the 'ats2-cuda-10.1.243-xl-2020.03.18_spmpi-rolling' builds on 'vortex'?

Yes, the error snip I posted was a sample from that build but I didn't indicate in the message (I'll update). For reference, the chama errors are much more cryptic but the underlying cause is the same problem due to use of an older gcc (I encountered these errors with older intel compilers), here's a sample snip

chama error sample
```
/gpfs1/jenkins/chama-slave/workspace/Trilinos-atdm-tlcc2-intel-debug-openmp/SRC_AND_BUILD/Trilinos/packages/kokkos/core/src/impl/Kokkos_EBO.hpp(128): error: invalid type for defaulted assignment operator
    EBOBaseImpl& operator=(EBOBaseImpl const&) = default;
```

**Comment by bartlettroscoe** [comment]
FYI: I sent out an email to the SPARC dev admins list and CCed @ndellingwood, @rppawlo, @crtrott, and @kddevin.  Hopefully the @trilinos/framework team can help lead the way to resolve this issue as part of the transition of the ATDM Trilinos configuration and builds to them.  (This is a great learning test case.)

**Comment by etphipp** [comment]
This is an issue that comes up all of the time when using Intel (and now I guess XL) compilers, since they rely on the GNU headers for most/all of their standard library support, and often the module for the compiler loads by default a pretty old version of GCC.  My experience has generally been that swapping the GCC module with a newer one works just fine, but someone would need to verify that by building the ATDM codes with the newer GCC compiler module.  If we hadn't reverted the commit to the ATDM builds, we could have tested this by simply modifying the ATDM build environment scripts to swap out the GCC module.

**Comment by bartlettroscoe** [comment]
> If we hadn't reverted the commit to the ATDM builds, we could have tested this by simply modifying the ATDM build environment scripts to swap out the GCC module.

@etphipp, we can still try this on this branch.  We just should not blow up the ATDM Trilinos nightly builds until we get this fixed.

**Comment by bartlettroscoe** [comment]
> My experience has generally been that swapping the GCC module with a newer one works just fine, but someone would need to verify that by building the ATDM codes with the newer GCC compiler module.

@etphipp, also, the question is if this will requiring rebuilding the TPL stack if we swap out the GCC backend.  We don't control the TPL stack on these machines, SPARC does.

Does someone on this thread have the time and interest to experiment with this on 'vortex' and 'tlcc2'?

**Comment by prwolfe** [comment]
> > My experience has generally been that swapping the GCC module with a newer one works just fine, but someone would need to verify that by building the ATDM codes with the newer GCC compiler module.
> 
> @etphipp, also, the question is if this will requiring rebuilding the TPL stack if we swap out the GCC backend. We don't control the TPL stack on these machines, SPARC does.
> 
> Does someone on this thread have the time and interest to experiment with this on 'vortex' and 'tlcc2'?

Ross - I looked at the tlcc2 build. I changed the environment.sh script to use gcc 7.2.0 instead of 4.9.3 to back up the intel compiler. The build has progressed well beyond the kokkos/core errors I was initially getting. I will re-check in the morning, but I think this will resolve the issue.

**Comment by bartlettroscoe** [comment]
> Ross - I looked at the tlcc2 build. I changed the environment.sh script to use gcc 7.2.0 instead of 4.9.3 to back up the intel compiler. The build has progressed well beyond the kokkos/core errors I was initially getting. I will re-check in the morning, but I think this will resolve the issue.

@prwolfe, thanks for doing that.  Can you post a PR for that change once you get confirmation that it is working?  We can then ask @jmgate to test this against EMPIRE if he wants (before this gets merged to the 'develop' branch).

Then we just need to figure out the XL compiler issue.

**Comment by prwolfe** [comment]
> > Ross - I looked at the tlcc2 build. I changed the environment.sh script to use gcc 7.2.0 instead of 4.9.3 to back up the intel compiler. The build has progressed well beyond the kokkos/core errors I was initially getting. I will re-check in the morning, but I think this will resolve the issue.
> 
> @prwolfe, thanks for doing that. Can you post a PR for that change once you get confirmation that it is working? We can then ask @jmgate to test this against EMPIRE if he wants (before this gets merged to the 'develop' branch).
> 
> Then we just need to figure out the XL compiler issue.

Ross - the build completed just fine, but the submission for the tests failed as it used acid's I am not on.

the change is minimal, 

`[prwolfe@chama-login3 Trilinos]$ git diff
diff --git a/cmake/std/atdm/common/toss3/environment_tlcc2.sh b/cmake/std/atdm/common/toss3/environment_tlcc2.sh
index 211aa5c..25f4194 100755
--- a/cmake/std/atdm/common/toss3/environment_tlcc2.sh
+++ b/cmake/std/atdm/common/toss3/environment_tlcc2.sh
@@ -47,7 +47,7 @@ if [ "$ATDM_CONFIG_COMPILER" == "INTEL" ]; then
     module load sems-boost/1.66.0/base
 
     module swap mkl/18.0.0.128 mkl/18.0.5.274
-    module load gcc/4.9.3
+    module load cde/compiler/gcc/7.2.0
     module unload sems-python/2.7.9
     export BOOST_ROOT=$SEMS_BOOST_ROOT
     export HDF5_ROOT=$SEMS_HDF5_ROOT
`

I would push it onto this branch unless you see a reason to do a different PR.

Paul

**Comment by prwolfe** [comment]
Well GitHub mangled that diff - sorry about that.

**Comment by jmgate** [comment]
Just go ahead and let it merge.  If it causes a problem for EMPIRE, we'll deal with it then.

**Comment by prwolfe** [comment]
The xlC compiles are not recognizing the -std=c++14 flag. The man page goes up only to c++1y. I will have to look at the nvcc_wrapper to see how that is getting decided.

File: cmake/std/atdm/common/toss3/environment_tlcc2.sh
Comment (bartlettroscoe) [inline cmake/std/atdm/common/toss3/environment_tlcc2.sh:50]: Wow, that easy?

Comment (etphipp) [reply cmake/std/atdm/common/toss3/environment_tlcc2.sh:50]: From my experience, yes it is that easy.

**Comment by prwolfe** [comment]
Seems so – I was surprised…

File: cmake/std/atdm/common/toss3/environment_tlcc2.sh
Comment (bartlettroscoe) [reply cmake/std/atdm/common/toss3/environment_tlcc2.sh:50]: @prwolfe, and @etphipp,

What impact, if any, would this have on the existing upstream TPLs that may have been built with GCC 4.9.3?  I guess if they don't use the C++ standard library, then it should be okay?  What about the standard C library (that TPLs like SuperLUDist must be using at least some)  with these two GCC versions.  Otherwise, are the header files and standard library in GCC 7.2.0 guaranteed to be backward compatible with the header files and standard library with GCC 4.9.3?

That is my main concern here.  I think we will need to build and run the full Trilinos test suite to be sure.  And someone really should be testing against SPARC and EMPIRE.  (That is what the ATDM DevOps team has done for 2+ years.)

**Comment by bartlettroscoe** [comment]
@jmgate, @prwolfe 

> Just go ahead and let it merge. If it causes a problem for EMPIRE, we'll deal with it then.

Just to be clear, that is okay for the 'tlcc2' builds for EMPIRE but we still need to address the issue with the XL compiler on ATS-2 'vortex' for SPARC.  Therefore, we can't merge this until we know that SPARC is going to be okay with CUDA+XL on ATS-2.

File: cmake/std/atdm/common/toss3/environment_tlcc2.sh
Comment (etphipp) [reply cmake/std/atdm/common/toss3/environment_tlcc2.sh:50]: That certainly can be an issue and this likely wouldn't work in all cases.  But to me the question is does it work in our case or not, which can easily be tested by just trying to compile and link the codes.  The ideal solution would be to recompile everything with gcc7 (or any gcc version where the c++14 support is not experimental), and that should be done long term even if swapping the compiler works in this case.

**Comment by crtrott** [comment]
XL magic: 
```
-DCMAKE_CXX_FLAGS="-isystem /home/projects/ppc64le/gcc/7.2.0/include/c++/7.2.0/powerpc64le-unknown-linux-gnu -qgcc_cpp_stdinc=/home/projects/ppc64le/gcc/7.2.0/include/c++/7.2.0 -qxflag=disable__cplusplusOverride"

```

**Comment by bartlettroscoe** [comment]
FYI: You can add arbitrary C++ flags for a build by setting the env var:

```
   export ATDM_CONFIG_CXX_FLAGS=”<your-flags>”
```

in the file:

```
   Trilinos/cmake/std/atdm/ats2/environment.sh 
```

likely around line:

https://github.com/trilinos/Trilinos/blob/a100d634cec2b9a4fd50660702996473ff18e49d/cmake/std/atdm/ats2/environment.sh#L109

You can see an example of setting C++ flags in the file:

```
    Trilinos/cmake/std/atdm/cee-rhel6/environment.sh
```

You can see how this is used in the Trilinos CMake configuration at:

https://github.com/trilinos/Trilinos/blob/a100d634cec2b9a4fd50660702996473ff18e49d/cmake/std/atdm/ATDMDevEnvSettings.cmake#L229

(Remember, most of this was copied from the EMPIRE configuration of Trilinos to match how they wanted to specify things.)

File: cmake/std/atdm/common/toss3/environment_tlcc2.sh
Comment (crtrott) [reply cmake/std/atdm/common/toss3/environment_tlcc2.sh:50]: We could be more conservative and swap for GCC 5.4 or so if that exists. But generally GCC versions are supposed to ABI compatible ...

**Comment by jwillenbring** [comment]
Does anything else need to be done for this? Can this be merged now?

**Comment by bartlettroscoe** [comment]
> Does anything else need to be done for this? Can this be merged now?

No, not unless you want to break the XL builds for SPARC on ATS-2.  See [above](https://github.com/trilinos/Trilinos/pull/8317#issuecomment-725692648).

Please don't knowingly break the ATDM Trilinos builds until these builds are fully handled over to the @trilinos/framework team.

**Comment by bartlettroscoe** [comment]
CC: @crtrott, @rppawlo, @etphipp, @trilinos/framework, @jwillenbring, @prwolfe 

Here is an email on this I sent out on Monday.

---

Sent: Monday, November 23, 2020 1:28 PM
To: Trott, Christian Robert
Cc: Glass, Micheal W ; Phipps, Eric T ; Pawlowski, Roger P
Subject: Trilinos and C++14

Hello Christian,

Responding to your comment about Trilinos and C++14 at the ATDM Weekly PI meeting today …

I created the PR:

   https://github.com/trilinos/Trilinos/pull/8317

back on Nov 5 to update Trilinos to require C++14 and did testing on all the ATDM platforms.  The XL compiler on ‘vortex’ and the Intel 18 compiler on TLCC-2 machines where flagged as non-C++14 compliant.  As you said, the problem is that the backend GCC compilers are too old (gcc/4.9.3).  Since C++14 is not required to complete the L1 milestone and since the ownership of the ATDM Trilinos configuration and builds is being transitioned to the Trilinos Framework team (and I am transitioning off), I have left this for someone else to resolve.  Paul Wolfenbarger (CCed) did a little work on this as commented in that PR but it is not clear where that effort stands. (There have been no new comments on that PR in 12 days.)  At this point, it is not clear who’s responsibility it is to fix these issues so that PR #8317 can be merged.  Otherwise, a new Kokkos version that requires C++14 can not be merged to Trilinos ‘develop’ until it is or it will break Trilinos for EMPIRE (on TLCC-2) and SPARC (on ATS-2 with XL).

Cheers,

-Ross

**Comment by prwolfe** [comment]
> XL magic:
> 
> ```
> -DCMAKE_CXX_FLAGS="-isystem /home/projects/ppc64le/gcc/7.2.0/include/c++/7.2.0/powerpc64le-unknown-linux-gnu -qgcc_cpp_stdinc=/home/projects/ppc64le/gcc/7.2.0/include/c++/7.2.0 -qxflag=disable__cplusplusOverride"
> ```
> 
> Oh and you need a fix in nvcc_wrapper to recognize -isystem ...

Can you give me a reference to this fix for mvcc_wrapper?

**Comment by bartlettroscoe** [comment]
@prwolfe, @crtrott 

FYI: Mike Glass says this fixes the problem with C++14 with XL no ATS-2:

```
export XLC_USR_CONFIG=/usr/tce/packages/xl/xl-2020.09.17/xlC/16.1.1/etc/xlc.cfg.rhel.7.6.gcc.8.3.1.cuda.10.1.v100
export XLF_USR_CONFIG=/usr/tce/packages/xl/xl-2020.09.17/xlf/16.1.1/etc/xlf.cfg.rhel.7.6.gcc.8.3.1.cuda.10.1.v100
```

This will cause the XL compiler to pick up and use gcc-8.3.1.

**Comment by crtrott** [comment]
Sounds good, the -isystem fix in nvcc wrapper is part of kokkos 3.3 otherwise, but i bet that config file contains that stuff

**Comment by prwolfe** [comment]
> Sounds good, the -isystem fix in nvcc wrapper is part of kokkos 3.3 otherwise, but i bet that config file contains that stuff

I can't find anything in that version of mvcc_wrapper that should handle -isystem. Cn you give me a commit to reference?

**Comment by bartlettroscoe** [comment]
@prwolfe, should be no need for any changes to `nvcc_wrapper` if you do what is called out [above](https://github.com/trilinos/Trilinos/pull/8317#issuecomment-738491977).

**Comment by crtrott** [comment]
Also the isystem addition is part of the release-candidate-3.3 branch of Kokkos, which is what we want to merge in. But as Ross said, just bend the config file for XL to the updated version. That is much preferrable over adding cxx flags.

**Comment by crtrott** [comment]
So we didn't end up adding the isystem thing. With a correct XL config file and adding "-ccbin xlc++ -std=c++14 -qxflag=disable__cplusplusOverride" to the CXX flags I can build NVCC+XL Trilinos (so far only had Tpetra enabled and Kokkos/KokkosKernels with tests) with the release candidate branch for Kokkos. #8490 

On white 
```
export XLF_USR_CONFIG=/opt/ibm/xlf/16.1.1/etc/xlf.cfg.rhel.7.5.gcc.7.2.0.cuda.10.1
export XLC_USR_CONFIG=/opt/ibm/xlC/16.1.1/etc/xlc.cfg.rhel.7.5.gcc.7.2.0.cuda.10.1
```

Similar config files should exist on the other machines. You just need one with a new enough GCC (5.4 or higher).

**Comment by prwolfe** [comment]
> So we didn't end up adding the isystem thing. With a correct XL config file and adding "-ccbin xlc++ -std=c++14 -qxflag=disable__cplusplusOverride" to the CXX flags I can build NVCC+XL Trilinos (so far only had Tpetra enabled and Kokkos/KokkosKernels with tests) with the release candidate branch for Kokkos. #8490

This may supersede the request I sent this morning. I will go work with this today. I may call you later if I get stuck.

Thanks!

**Comment by crtrott** [comment]
I checked into Vortex itself to see if I can get it build and looks like I am successful. Unfortunately the existing config files don't work. Apparently XL doesn't like GCC 8.3.0 ...

So what I did is generate myself a config file with gcc 7. These config files fortunately record how they were generated so that was reasonably straight forward:

```
/usr/tce/packages/xl/xl-2020.03.18/xlC/16.1.1/bin/new_xlc_configure -ibmcmp /usr/tce/packages/xl/xl-2020.03.18 -cuda /usr/tce/packages/cuda/cuda-10.1.243 -cudaVersion 10.1.243 -cuda_cc_major 7 -cuda_cc_minor 0 -gcc /usr/tce/packages/gcc/gcc-7.3.1 -o ~/xlc.cfg.gcc7.3.1
```

Module list:
```
Currently Loaded Modules:
  1) StdEnv (S)   2) python/2.7.16   3) sparc-tools/aerotools/2   4) sparc-tools/taos/2020.09.04   5) xl/2020.03.18   6) spectrum-mpi/rolling-release   7) cuda/10.1.243   8) lapack/3.8.0-xl-2019.08.20   9) sparc-dev/cuda-10.1.243_xl-2020.03.18_spmpi-rolling  10) git/2.20.0  11) cmake/3.18.0
```

Mostly I loaded the standard environment via:
```
source ~/trilinos/cmake/std/atdm/load-env.sh Volta70-cuda-xl-opt-shared
```

But you also need to swap the cmake, since a newer cmake is needed so that it knows that NVCC can deal with C++14.

My configure:
```
#!/bin/bash
DISABLE_LONG_PANZER_TESTS=OFF

TRILINOS_DEBUG=OFF
if [[ "${ATDM_CONFIG_BUILD_TYPE}" == "DEBUG" ]]; then
  DISABLE_LONG_PANZER_TESTS=ON
  TRILINOS_DEBUG=ON
fi

useNinja=""
if [[ "${BUILD_COMMAND}" == "ninja" ]]; then
  useNinja="-G Ninja"
fi

if [[ "${INSTALL_DIR}" == "" ]]; then
  echo "No install directory set, exiting before running cmake"
  exit -1
fi

cmake \
  ${useNinja} \
  -D ATDM_TRILINOS_INSTALL_PREFIX_ENV_VAR_NAME=Trilinos_INSTALL_DIR \
  -D BUILD_SHARED_LIBS:BOOL=${ATDM_CONFIG_SHARED_LIBS} \
  -D CMAKE_INSTALL_PREFIX:FILEPATH=${INSTALL_DIR} \
  -D Trilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake \
  -D Trilinos_ENABLE_Tpetra:BOOL=ON \
  -D TPL_ENABLE_gtest=OFF \
  -D Trilinos_ENABLE_Gtest=OFF \
  -D Trilinos_ENABLE_DEBUG=${TRILINOS_DEBUG} \
  -D CMAKE_CXX_STANDARD=14 \
  -D CMAKE_CXX_FLAGS="-ccbin xlc++ -std=c++14 -qxflag=disable__cplusplusOverride" \
  ${HOME}/trilinos
```

**Comment by crtrott** [comment]
Tpetra did build successfully.

**Comment by crtrott** [comment]
@prwolfe did you have a chance to verify this?

**Comment by prwolfe** [comment]
> @prwolfe did you have a chance to verify this?

I confirmed what you saw with gcc 8.3 and have not yet set up a gcc 7 file. We will need to get that installed in a common location - can you request that from the admins? If they are not amenable we can find a place to share that among Trilinos, SPARC, and Kokkos (yeah - that's messy at best...)

**Comment by bartlettroscoe** [comment]
Remember,  whatever gets installed and used has to work on the production ATS-2 env as well.

**Comment by prwolfe** [comment]
Talked with Kevin Buterbaugh this afternoon and he will be talking with the Vortex admin team about generating the config file and placing it in /opt/ibm/xlC/16.1.1/etc/xlc.cfg.rhel.7.5.gcc.7.3.1.cuda.10.1.243. This is not an area synced from SIERRA where /usr/tce is and he expects they can add this to an existing post-processing script. I will follow up with him next week.

**Comment by prwolfe** [comment]
Kevin got agreement from the rest of the admin team and has installed the xlC config file. As soon as the elf file is in place I will create the needed changes and test the branch on vortex.

**Review by crtrott — APPROVED** [review]

**Comment by crtrott** [comment]
Are we good to go on this and deal with Vortex XL builds separately? Or do you want to fix the XL stuff as part of this? This kinda blocks #8490  @bartlettroscoe

**Comment by crtrott** [comment]
actually someone probably needs to fix the conflicts first.

**Comment by bartlettroscoe** [comment]
@crtrott, how can you approve this if we don't even know if this works yet?

**Comment by crtrott** [comment]
My understanding from this is that we do know it works, with the exception of Vortex, which Paul said in a call we might want to fix separately.

**Comment by prwolfe** [comment]
> My understanding from this is that we do know it works, with the exception of Vortex, which Paul said in a call we might want to fix separately.

I am very close to sending these - basically it works with a few minor changes
* had to move forward to cmake 3.18, which I did not expect as the branch should have been self-consistent
* Had to use the config file I generated as the one the admins have installed still references 8.3.1 (I sent them a ticket)

So the real delay here is to get well-supported config files.

I will look at resolving the conflicts in this branch and getting it updated.

**Review by jwillenbring — APPROVED** [review]
I reviewed the most recent change. Looks good. @crtrott previously reviewed other commits.

**Comment by crtrott** [comment]
There is still a conflict.

**Comment by jwillenbring** [comment]
Closing this PR in favor of #8642

**Comment by bartlettroscoe** [comment]
Related to:
* SEPW-213

**Pull request closed without merging**
